README.md
# hubot-dokku-hook

[![Greenkeeper badge](https://badges.greenkeeper.io/lgaticaq/hubot-dokku-hook.svg)](https://greenkeeper.io/)

[![npm version](https://img.shields.io/npm/v/hubot-dokku-hook.svg?style=flat-square)](https://www.npmjs.com/package/hubot-dokku-hook)
[![npm downloads](https://img.shields.io/npm/dm/hubot-dokku-hook.svg?style=flat-square)](https://www.npmjs.com/package/hubot-dokku-hook)
[![Build Status](https://img.shields.io/travis/lgaticaq/hubot-dokku-hook.svg?style=flat-square)](https://dokku.org/lgaticaq/hubot-dokku-hook)
[![Coverage Status](https://img.shields.io/coveralls/lgaticaq/hubot-dokku-hook/master.svg?style=flat-square)](https://coveralls.io/github/lgaticaq/hubot-dokku-hook?branch=master)
[![Code Climate](https://img.shields.io/codeclimate/github/lgaticaq/hubot-dokku-hook.svg?style=flat-square)](https://codeclimate.com/github/lgaticaq/hubot-dokku-hook)
[![dependency Status](https://img.shields.io/david/lgaticaq/hubot-dokku-hook.svg?style=flat-square)](https://david-dm.org/lgaticaq/hubot-dokku-hook#info=dependencies)
[![devDependency Status](https://img.shields.io/david/dev/lgaticaq/hubot-dokku-hook.svg?style=flat-square)](https://david-dm.org/lgaticaq/hubot-dokku-hook#info=devDependencies)

> Get Dokku deploy notification

## Install

```bash
npm i -S hubot-dokku-hook
```

Add `["hubot-dokku-hook"]` in `external-scripts.json`.

Install [dokku-webhooks](https://github.com/nickstenning/dokku-webhooks) in your dokku and set the webhook with:

```bash
dokku webhooks:add http://hubot.example.com/dokku/room-name
```

Replace `room-name` with the slack or hubot channel

![notification](http://pix.toile-libre.org/upload/original/1489525873.png)

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
## 1.0.0 (2018-02-17)

##### Chores

* **deps:**
  *  update dependency coffeelint to v2.1.0 ([d774bf33](https://github.com/lgaticaq/hubot-dokku-hook/commit/d774bf33e5d4bea3c491f0e7135b8ab4937ca7fa))
  *  update dependency mocha to v5.0.1 ([ce3ec03e](https://github.com/lgaticaq/hubot-dokku-hook/commit/ce3ec03e37de110e2acfb69de5fd23ddbd0ecf5f))
  *  update dependency mocha to v5.0.0 ([098e5cc0](https://github.com/lgaticaq/hubot-dokku-hook/commit/098e5cc0847f3820370f9975efa750a40b26a623))
  *  update dependency mocha to v4.1.0 ([809d5364](https://github.com/lgaticaq/hubot-dokku-hook/commit/809d5364973ac9183985627d2057260cb28b0213))
  *  update dependency generate-changelog to v1.7.0 ([3ffd865b](https://github.com/lgaticaq/hubot-dokku-hook/commit/3ffd865b37f9c4a6328fee3d717bd6bab0ed22ff))
  *  update dependency coffee-script to v1.12.8 ([ff4a5379](https://github.com/lgaticaq/hubot-dokku-hook/commit/ff4a537989b53d0c0318d4ac905fb7e40688e234))
  *  update dependency generate-changelog to v1.6.0 ([d42577ee](https://github.com/lgaticaq/hubot-dokku-hook/commit/d42577eed0c41f850fa036e1a868234e042ab4ff))
  *  update dependency coffeelint to v2.0.7 ([e636fa75](https://github.com/lgaticaq/hubot-dokku-hook/commit/e636fa75aa8f03cfb3096a0c438c3f5a0f639a99))
  *  update dependency hubot-test-helper to v1.8.1 ([854a4702](https://github.com/lgaticaq/hubot-dokku-hook/commit/854a4702ab3ba3711173f50b5f5cbfd6c23555e2))
  *  update dependency hubot-test-helper to v1.8.0 ([07577e6f](https://github.com/lgaticaq/hubot-dokku-hook/commit/07577e6f237b080530c9d973b38d8a6f86f4dedf))
  *  update dependency mocha to v4.0.1 ([bc5dbfec](https://github.com/lgaticaq/hubot-dokku-hook/commit/bc5dbfec92170747d44f2f851c2705e38f4b5424))
  *  update dependency coffeelint to v2.0.6 ([6812fcc6](https://github.com/lgaticaq/hubot-dokku-hook/commit/6812fcc612958df7133255b76fd8389d11691200))
  *  update dependency coffeelint to v2.0.5 ([1c4c52ad](https://github.com/lgaticaq/hubot-dokku-hook/commit/1c4c52ad2b7b4d00406014a5bca254f077e23aac))
  *  update dependency coffeelint to v2.0.3 ([2ad46b45](https://github.com/lgaticaq/hubot-dokku-hook/commit/2ad46b459451f99737b36d5bbdc9b565ec547b80))
  *  update dependency mocha to v4.0.0 ([9e7bcb5c](https://github.com/lgaticaq/hubot-dokku-hook/commit/9e7bcb5cd79420f0f8dd9275f65ad707c5fd0856))
  *  update dependency coveralls to v3.0.0 ([f74ca95f](https://github.com/lgaticaq/hubot-dokku-hook/commit/f74ca95f8e64f1f473682f6bd74636ef009861ce))
  *  update dependency generate-changelog to v1.5.0 ([17df59f4](https://github.com/lgaticaq/hubot-dokku-hook/commit/17df59f4caf1e617a1a095f1523184cb939cfc5c))
  *  update dependency mocha to v3.5.3 ([db793551](https://github.com/lgaticaq/hubot-dokku-hook/commit/db793551b7fbc1296fdfd2d8ce0ff5957cc94eb4))
  *  update dependency mocha to v3.5.2 ([599f27d2](https://github.com/lgaticaq/hubot-dokku-hook/commit/599f27d25aff8dc67de229544c166e45a0b6636f))
* **package:**
  *  update mocha to version 3.5.3 ([af28de50](https://github.com/lgaticaq/hubot-dokku-hook/commit/af28de50ef71375106aad1acd21b9e9dd0cbbf5c))
  *  update codeclimate-test-reporter to version 0.5.0 ([c1c6d19e](https://github.com/lgaticaq/hubot-dokku-hook/commit/c1c6d19e21882fc2e750a496b2a10cc19e5c67ab))
  *  update chai to version 4.0.0 ([ea7bb239](https://github.com/lgaticaq/hubot-dokku-hook/commit/ea7bb2396109baff4ec3babe48f259553f293a02))

##### Documentation Changes

* **readme:**  add Greenkeeper badge ([7e6c4102](https://github.com/lgaticaq/hubot-dokku-hook/commit/7e6c4102bdaa61f0b4b1aee4633876e05e140ced))

##### New Features

* **js:**  migrate to js standard ([5d6fcc27](https://github.com/lgaticaq/hubot-dokku-hook/commit/5d6fcc27d01fb69b7271dffef8e1172af80db081))

##### Bug Fixes

* **deps:**  update dependency coffee-script to v1.12.7 ([e83e7d16](https://github.com/lgaticaq/hubot-dokku-hook/commit/e83e7d16f5ca58e7cc4e32acc3dd48f898c1c284))

##### Refactors

* **deps:**
  *  renovate pin dependencies packages ([d343ee1a](https://github.com/lgaticaq/hubot-dokku-hook/commit/d343ee1af3027bbec077e48ad9b7cb5d607dde00))
  *  renovate pin dependencies packages ([467fe47d](https://github.com/lgaticaq/hubot-dokku-hook/commit/467fe47db3583532a0912b3315d00b4e59957633))

#### 0.0.1 (2017-3-14)

##### Chores

* **project:** first commit ([3c366790](https://github.com/lgaticaq/hubot-dokku-hook/commit/3c3667905f536dc1e5918d66fee0d2f37ba43ab9))



test/test.js
'use strict'

const { describe, it, beforeEach, afterEach } = require('mocha')
const Helper = require('hubot-test-helper')
const { expect } = require('chai')
const http = require('http')
const querystring = require('querystring')

const helper = new Helper('./../src/index.js')

describe('travis-ci', function () {
  beforeEach(() => {
    this.room = helper.createRoom({ name: 'random' })
  })

  afterEach(() => {
    this.room.destroy()
  })

  describe('POST /dokku/random', () => {
    beforeEach(done => {
      const postData = querystring.stringify({
        action: 'post-deploy',
        app: 'myapp',
        url: 'http://myapp.example.com'
      })
      const postOptions = {
        hostname: 'localhost',
        port: 8080,
        path: '/dokku/random',
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Content-Length': Buffer.byteLength(postData)
        }
      }
      this.room.robot.adapter.client = {
        web: {
          chat: {
            postMessage: (channel, text, options) => {
              this.postMessage = {
                channel,
                text,
                options
              }
              done()
            }
          }
        }
      }
      const req = http.request(postOptions, response => {
        this.response = response
      })
      req.write(postData)
      req.end()
    })

    it('responds with status 200 and results', () => {
      expect(this.postMessage.channel).to.eql('#random')
      expect(this.postMessage.text).to.be.a('null')
      expect(this.postMessage.options.as_user).to.eql(false)
      expect(this.postMessage.options.link_names).to.eql(1)
      expect(this.postMessage.options.username).to.eql('Dokku')
      expect(this.postMessage.options.icon_url).to.eql(
        'https://avatars1.githubusercontent.com/u/13455795?v=3&s=200'
      )
      expect(this.postMessage.options.attachments).to.eql([
        {
          color: 'good',
          fallback: 'myapp deployed, check http://myapp.example.com',
          text: 'myapp deployed, check http://myapp.example.com'
        }
      ])
    })
  })
})


src/script.js
// Description
//   Get Dokku deploy notification
//
// Dependencies:
//   None
//
// Configuration:
//   None
//
// Commands:
//   None
//
// Author:
//   lgaticaq

'use strict'

const icon = 'https://avatars1.githubusercontent.com/u/13455795?v=3&s=200'

module.exports = robot => {
  robot.router.post('/dokku/:room', (req, res) => {
    const channel = req.params.room
    res.send('Ok')
    if (req.body.action === 'post-deploy') {
      const message = `${req.body.app} deployed, check ${req.body.url}`
      const attachments = {
        as_user: false,
        link_names: 1,
        icon_url: icon,
        username: 'Dokku',
        unfurl_links: false,
        attachments: [
          {
            fallback: message,
            color: 'good',
            text: message
          }
        ]
      }
      if (['SlackBot', 'Room'].includes(robot.adapter.constructor.name)) {
        return robot.adapter.client.web.chat.postMessage(
          `#${channel}`,
          null,
          attachments
        )
      }
      robot.messageRoom(channel, message)
    }
  })
}
